File: redo/osxaltest/window.m
// 1 august 2015
#include "osxaltest.h"

@implementation NSView (AutoLayoutHelpers)
- (void)tIsAmbiguous:(uintmax_t)indent
{
	NSMutableString *s;
	uintmax_t i;
	NSUInteger j;

	s = [NSMutableString new];
	for (i = 0; i < indent; i++)
		[s appendString:@" "];
	NSLog(@"%@%@ %d", s, [self className], (int) [self hasAmbiguousLayout]);
	if ([self hasAmbiguousLayout])
		[[self window] visualizeConstraints:[[self superview] constraints]];
	for (j = 0; j < [[self subviews] count]; j++)
		[[[self subviews] objectAtIndex:j] tIsAmbiguous:(indent + 1)];
}
@end

@implementation tWindow {
	NSWindow *w;
	id<tControl> c;
	BOOL margined;
}

- (id)init
{
	self = [super init];
	if (self) {
		self->w = [[NSWindow alloc] initWithContentRect:NSMakeRect(0, 0, 320, 240)
			styleMask:(NSTitledWindowMask | NSClosableWindowMask | NSMiniaturizableWindowMask | NSResizableWindowMask)
			backing:NSBackingStoreBuffered
			defer:YES];
		[self->w setTitle:@"Auto Layout Test"];
	}
	return self;
}

- (void)tSetControl:(id<tControl>)cc
{
	self->c = cc;
	[self->c tSetParent:self addToView:[self->w contentView] relayout:NO];
	[self tRelayout];
}

- (void)tSetMargined:(BOOL)m
{
	self->margined = m;
	[self tRelayout];
}

- (void)tShow
{
	[self->w cascadeTopLeftFromPoint:NSMakePoint(20, 20)];
	[self->w makeKeyAndOrderFront:self];
//	[[self->w contentView] tIsAmbiguous:0];
}

- (void)tRelayout
{
	NSView *contentView;
	tAutoLayoutParams p;
	NSDictionary *views;
	NSString *margin;
	NSMutableString *constraint;
	NSArray *constraints;

	if (self->c == nil)
		return;

	contentView = [self->w contentView];
	[contentView removeConstraints:[contentView constraints]];

	[self->c tFillAutoLayout:&p];

	views = [NSDictionary dictionaryWithObject:p.view forKey:@"view"];

	margin = @"";
	if (self->margined)
		margin = @"-";

	// TODO always append margins even if not attached?
	// or if not attached, append ->=0- as well?
	constraint = [NSMutableString stringWithString:@"H:"];
	if (p.attachLeft) {
		[constraint appendString:@"|"];
		[constraint appendString:margin];
	}
	[constraint appendString:@"[view]"];
	if (p.attachRight) {
		[constraint appendString:margin];
		[constraint appendString:@"|"];
	}
	constraints = [NSLayoutConstraint constraintsWithVisualFormat:constraint options:0 metrics:nil views:views];
	[contentView addConstraints:constraints];
	// TODO do not release constraint; it's autoreleased?

	constraint = [NSMutableString stringWithString:@"V:"];
	if (p.attachTop) {
		[constraint appendString:@"|"];
		[constraint appendString:margin];
	}
	[constraint appendString:@"[view]"];
	if (p.attachBottom) {
		[constraint appendString:margin];
		[constraint appendString:@"|"];
	}
	constraints = [NSLayoutConstraint constraintsWithVisualFormat:constraint options:0 metrics:nil views:views];
	[contentView addConstraints:constraints];
	// TODO do not release constraint; it's autoreleased?

	// TODO do not release views; it's autoreleased?
	// (for all of these, look up whether the constructor autoreleases)
}

@end

File: redo/osxaltest/window.swift
// 1 august 2015
import Cocoa

// auto layout helpers
func isAmbiguous(view: NSView, indent: Int) {
	var s = String(count: indent, repeatedValue: " " as Character)
	println("\(s) \(view.className) \(view.hasAmbiguousLayout)")
	if view.hasAmbiguousLayout {
		view.window?.visualizeConstraints(view.superview!.constraints)
	}
	for subview in view.subviews {
		isAmbiguous(subview as! NSView, indent + 1)
	}
}

class Window : NSWindow, Control {
	private var c: Control?
	private var margined: Bool

	init() {
		self.c = nil
		self.margined = false
		// we have to initialize our own instance variables first, unfortunately (thanks erica in irc.freenode.net/#swift-lang)
		super.init(
			contentRect: NSMakeRect(0, 0, 320, 240),
			styleMask: (NSTitledWindowMask | NSClosableWindowMask | NSMiniaturizableWindowMask | NSResizableWindowMask),
			backing: NSBackingStoreType.Buffered,
			defer: true)
		self.title = "Auto Layout Test"
	}

	required init?(coder: NSCoder) {
		fatalError("can't use this constructor, sorry")
	}

	func SetControl(c: Control) {
		self.c = c
		var contentView = self.contentView as! NSView
		contentView.addSubview(self.c!.View())
		self.relayout()
	}

	func SetMargined(m: Bool) {
		self.margined = m
		self.relayout()
	}

	func Show() {
		self.cascadeTopLeftFromPoint(NSMakePoint(20, 20))
		self.makeKeyAndOrderFront(self)
		isAmbiguous(self.contentView as! NSView, 0)
	}

	func View() -> NSView {
		fatalError("cannot call Window.View()")
	}

	func SetParent(p: Control) {
		fatalError("cannot call Window.SetParent()")
	}

	private func relayout() {
		if self.c == nil {
			return
		}

		var contentView = self.contentView as! NSView
		contentView.removeConstraints(contentView.constraints)

		var views = [
			"view":	self.c!.View(),
		]
		var margin = ""
		if self.margined {
			margin = "-"
		}

		var constraint = "H:|" + margin + "[view]" + margin + "|"
		var constraints = mkconstraints(constraint, nil, views)
		contentView.addConstraints(constraints)

		constraint = "V:|" + margin + "[view]" + margin + "|"
		constraints = mkconstraints(constraint, nil, views)
		contentView.addConstraints(constraints)
	}
}
